### PR TITLE
Avoid allocating a vector with typemax(Int) entries from read

### LINKED ISSUE #30978: read(io, typemax(Int)) throws OutOfMemoryError, read(io) does not
from the help:

```
read(s::IO, nb=typemax(Int))
Read at most nb bytes from s, returning a Vector{UInt8} of the bytes read.
```

but passing typemax(Int) as second argument actually tries to allocate that memory and fails. So the one-argument reads does something other than advertised in the help.

I would like to use typemax(Int) as a default argument for my own function, and pass it to read. read(s, nb) throws an OutOfMemoryError, so I have to test and use the 1-arg read(s) instead.

### PR DESCRIPTION
**Pull request by nalimilan**
Also fix a bug when `all=false` and the input vector is too small, and improve docstrings a bit.

Fixes #30978.

### CODE REVIEW

File: test/iostream.jl [hunk 1]
```diff
@@ -83,3 +83,20 @@ end
 end
 
 @test Base.open_flags(read=false, write=true, append=false) == (read=false, write=true, create=true, truncate=true, append=false)
+
+@testset "issue #30978" begin
+    path, io = mktemp()
```
Comment (JeffBezanson) [inline test/iostream.jl]: Should use `mktemp() do ...` to ensure the file is deleted.

File: base/iostream.jl [hunk 2]
```diff
@@ -458,14 +458,15 @@ function readbytes_all!(s::IOStream, b::Array{UInt8}, nb)
...
 function readbytes_some!(s::IOStream, b::Array{UInt8}, nb)
-    olb = lb = length(b)
-    if nb > lb
+    olb = length(b)
+    if nb > olb
         resize!(b, nb)
     end
     nr = GC.@preserve b Int(ccall(:ios_read, Csize_t, (Ptr{Cvoid}, Ptr{Cvoid}, Csize_t),
                                   s.ios, pointer(b), nb))
+    lb = length(b)
```
Comment (JeffBezanson) [inline base/iostream.jl]: :+1: Thanks!

**Comment by JeffBezanson** [comment]
I guess this is ok for now, but I really don't want this magic value to proliferate. It's not a great API, and we can't promise that it's generally safe to ask for `typemax(Int)` bytes, expecting every read method to handle it.

**Comment by nalimilan** [comment]
OK. Do you think we should switch to `nothing` then?

I've pushed another commit fixing another corner case when `all=false`.

**Comment by nalimilan** [comment]
Good to go?

**Pull request merged**